src/shogun/classifier/NearestCentroid.cpp

```diff
@@ -54,51 +53,41 @@ namespace shogun{
 		{
 			data = distance->get_lhs();
 		}
-		int32_t num_vectors = data->get_num_vectors();
-		int32_t num_classes = ((CMulticlassLabels*) m_labels)->get_num_classes();
-		int32_t num_feats = ((CDenseFeatures<float64_t>*) data)->get_num_features();
-		SGMatrix<float64_t> centroids(num_feats,num_classes);
-		centroids.zero();
 
-		int64_t* num_per_class = new int64_t[num_classes];
-		for (int32_t i=0 ; i<num_classes ; i++)
-		{
-			num_per_class[i]=0;
-		}
+		auto* multiclass_labels = m_labels->as<CMulticlassLabels>();
+		auto* dense_data = data->as<CDenseFeatures<float64_t>>();
+
+		int32_t num_classes = multiclass_labels->get_num_classes();
+		int32_t num_feats = dense_data->get_num_features();
 
-		for (int32_t idx=0 ; idx<num_vectors ; idx++)
+		SGMatrix<float64_t> centroids(num_feats, num_classes);
+		SGVector<int64_t> num_per_class(num_classes);
+
+		auto iter_labels = multiclass_labels->get_int_labels().begin();
```

Comment:
This generates a copy I think. Is that necessary? We can also cast the labels on the fly below.

Also. This reminds me of another thing. I don't like this method, or any of the other "casting" methods in `CDenseLabels`. A good clean up PR would be

* remove all those methods and only keep `get_labels()` and `get_label(index_t)`.
* Instead, add a casting method `SGVector<T>::as<S>()` which returns a new vector with the converted data type
* Replace calls `labels->get_int_labels()` with the more explicit `labels->get_labels()->as<int32_t>()`

Reply:
`SGVector<T>::as<S>` is already implemented. Should I use it here? It creates a copy as well.